pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/funnel/FunnelMaxStepAggregationFunction.java

```diff
@@ -0,0 +1,302 @@
+    while (!stepEvents.isEmpty()) {
+      fillWindow(stepEvents, slidingWindow);
+      if (slidingWindow.isEmpty()) {
+        break;
+      }
+      int maxSteps = processWindow(slidingWindow);
+      finalMaxStep = Math.max(finalMaxStep, maxSteps);
+      if (finalMaxStep == _numSteps) {
+        break;
+      }
+      long windowStart = slidingWindow.peek().getTimestamp();
+      while (!slidingWindow.isEmpty() && slidingWindow.peek().getTimestamp() == windowStart) {
+        slidingWindow.pollFirst();
+      }
+    }
+    return finalMaxStep;
+  }
+
+  private void fillWindow(PriorityQueue<FunnelStepEvent> stepEvents, ArrayDeque<FunnelStepEvent> slidingWindow) {
+    // Ensure for the sliding window, the first event is the first step
+    while ((!slidingWindow.isEmpty()) && slidingWindow.peek().getStep() != 0) {
+      slidingWindow.pollFirst();
+    }
+    if (slidingWindow.isEmpty()) {
+      while (!stepEvents.isEmpty() && stepEvents.peek().getStep() != 0) {
+        stepEvents.poll();
+      }
+      if (stepEvents.isEmpty()) {
+        return;
+      }
+      slidingWindow.addLast(stepEvents.poll());
+    }
+    // SlidingWindow is not empty
+    long windowStart = slidingWindow.peek().getTimestamp();
+    while (!stepEvents.isEmpty() && (stepEvents.peek().getTimestamp() <= windowStart + _windowSize)) {
+      slidingWindow.addLast(stepEvents.poll());
+    }
+  }
+
+  private int processWindow(ArrayDeque<FunnelStepEvent> slidingWindow) {
+    int maxStep = 0;
+    long previousTimestamp = -1;
+    for (FunnelStepEvent event : slidingWindow) {
+      int currentEventStep = event.getStep();
+      // If the same condition holds for the sequence of events, then such repeating event interrupts further
+      // processing.
+      if (_modes.contains(Mode.strict_deduplication)) {
+        if (currentEventStep == maxStep - 1) {
+          return maxStep;
+        }
+      }
+      // Don't allow interventions of other events. E.g. in the case of A->B->D->C, it stops finding A->B->C at the D
+      // and the max event level is 2.
+      if (_modes.contains(Mode.strict_order)) {
+        if (currentEventStep != maxStep) {
+          return maxStep;
+        }
+      }
+      // Apply conditions only to events with strictly increasing timestamps.
+      if (_modes.contains(Mode.strict_increase)) {
+        if (previousTimestamp == event.getTimestamp()) {
+          continue;
+        }
+      }
+      previousTimestamp = event.getTimestamp();
+      if (maxStep == currentEventStep) {
+        maxStep++;
+      }
+      if (maxStep == _numSteps) {
+        break;
+      }
+    }
+    return maxStep;
+  }
+
+  @Override
+  public String toExplainString() {
+    return "WindowFunnelAggregationFunction{"
+        + "_timestampExpression=" + _timestampExpression
+        + ", _windowSize=" + _windowSize
+        + ", _stepExpressions=" + _stepExpressions
+        + ", _numSteps=" + _numSteps
+        + '}';
+  }
+
+  enum Mode {
```

Comment:
Is this private access only?
We usually define enum using all capital letters. When reading it, we should convert value to upper case to match different cases

Reply:
done.